modules/saml/src/Controller/ServiceProvider.php

```diff
@@ -98,36 +102,110 @@ public function setAuthUtils(Utils\Auth $authUtils): void
+            throw new Error\Exception('Authsource must be of type saml:SP.');
+        }
+
+        $httpUtils = new Utils\HTTP();
+
+        $returnTo = $this->loginHandler($request, $authSource,  $spSource, $httpUtils);
+
+        // Redirect to the returnTo destination
+        return new RunnableResponse([$httpUtils, 'redirectTrustedURL'], [$returnTo]);
+    }
+
     /**
-     * Start single sign-on for an SP identified with the specified Authsource ID
+     * @param   Request  $request
+     * @param   Simple   $authSource
+     * @param   Source   $spSource
+     * @param   HTTP     $httpUtils
      *
-     * @param \Symfony\Component\HttpFoundation\Request $request
-     * @param string $sourceId
-     * @return \SimpleSAML\HTTP\RunnableResponse
+     * @return string
+     * @throws BadRequest
+     * @throws Error\Exception
      */
-    public function login(Request $request, string $sourceId): RunnableResponse
+    protected function loginHandler(Request $request,
+                                    Auth\Simple $authSource,
+                                    Auth\Source $spSource,
+                                    Utils\HTTP $httpUtils): string
     {
-        $as = new Auth\Simple($sourceId);
-        if (!($as->getAuthSource() instanceof SP)) {
-            throw new Error\Exception('Authsource must be of type saml:SP.');
+        $options = [];
+
+        if ($spSource->isRequestInitiation()) {
+            if ($request->query->has('target')) {
+                $options['ReturnTo'] = $httpUtils->checkURLAllowed($request->query->get('target'));
+            }
+            if ($request->query->has('forceAuthn')) {
+                $options['ForceAuthn'] = $request->query->get('forceAuthn');
+            }
+            if ($request->query->has('entityID')) {
+                $options['saml:idp'] = $request->query->get('entityID');
+            }
+            if ($request->query->has('isPassive')) {
+                $options['isPassive'] = $request->query->get('isPassive');
+            }
         }
 
-        if (!$request->query->has('ReturnTo')) {
+        if (
+            !isset($options['ReturnTo'])
+            && !$request->query->has('ReturnTo')
+            && $spSource->getMetadata()->getOptionalString('RelayState', null) === null
+        ) {
             throw new Error\BadRequest('Missing ReturnTo parameter.');
         }
-        $returnTo = $request->query->get('ReturnTo');
+        if (!isset($options['ReturnTo'])) {
+            $options['ReturnTo'] = $httpUtils->checkURLAllowed(
+                $request->query->get('ReturnTo') ?? $spSource->getMetadata()->getOptionalString('RelayState', null)
+            );
+        }
+
+        $authData = $authSource->getAuthDataArray();
+        // We are copying the current IDP to the targeted one in case we want to ForceAuthn
+        if(empty($options['saml:idp']) && !empty($authData['saml:sp:IdP'])) {
+            $options['saml:idp'] = $authData['saml:sp:IdP'];
```

Comment:
I don't think this is true. We may be using ForceAuthn to send the user to discovery. I'd remove this part.

Reply:
Done